pkg/cli/cluster_free_disk_space.go

```diff
@@ -0,0 +1,125 @@
+package cli
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+
+	"github.com/spf13/cobra"
+	v1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	rookcli "github.com/rook/rook/pkg/client/clientset/versioned"
+
+	clusterspace "github.com/replicatedhq/kurl/pkg/cluster/space"
+)
+
+const (
+	IsDefaultStorageClassAnnotation = "storageclass.kubernetes.io/is-default-class"
+	OpenEBSLocalProvisioner         = "openebs.io/local"
+	RookRBDProvisioner              = "rook-ceph.rbd.csi.ceph.com"
+	RookCephFSProvisioner           = "rook-ceph.cephfs.csi.ceph.com"
+)
+
+func NewClusterFreeDiskSpaceCmd(cli CLI) *cobra.Command {
+	var forStorageClass, openEBSImage, openEBSNode string
+	logbuf := bytes.NewBuffer(nil)
+	logger := log.New(logbuf, "", 0)
+	k8sConfig := config.GetConfigOrDie()
+	clientSet := kubernetes.NewForConfigOrDie(k8sConfig)
+	rookClientSet := rookcli.NewForConfigOrDie(k8sConfig)
+
+	cmd := &cobra.Command{
+		Use:          "free-disk-space",
+		Short:        "Returns the amount of free disk space (in bytes) for a given storage class",
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			classes, err := clientSet.StorageV1().StorageClasses().List(cmd.Context(), metav1.ListOptions{})
```

Comment:
Could we properly encapsulate the methods into validate() and then run()?

Reply:
Could you please point me to an example of this ?

Reply:
I thought in something like: 


```go
// NewRookFreeDiskSpaceGetter returns a command that is capable of reporting back the amount of free space in the cluster
// for a provided storage class.
func NewClusterFreeDiskSpaceCmd(cli CLI) *cobra.Command {
	var forStorageClass, openEBSImage, openEBSNode string
	logger = log.New(logbuf, "", 0)

	cmd := &cobra.Command{
		Use:          "free-disk-space",
		Short:        "Returns the amount of free disk space (in bytes) for a given storage class",
		SilenceUsage: true,
		PreRunE: func(cmd *cobra.Command, args []string) (err error) {
			k8sConfig = config.GetConfigOrDie()
			clientSet = kubernetes.NewForConfigOrDie(k8sConfig)
			rookClientSet = rookcli.NewForConfigOrDie(k8sConfig)
		},
		RunE: func(cmd *cobra.Command, args []string) (err error) {
			// See that here we can do what is requited to be done
			defaultClassProvisioner,err:= findDefaultClass()
			if err != nil {
				return err
			}
			output,err:= runCheck(defaultClassProvisioner);
			if err != nil {
				return err
			}

			// OUTPUT the result
			fmt.Sprintf("the availble free disk space with thr provisioner %s is: (%s)",
				output,defaultClassProvisioner.Name)
			return nil
		},
	}

	cmd.Flags().StringVar(&forStorageClass, "storageclass", "", "The storage class for which to check the free disk space.")
	cmd.Flags().StringVar(&openEBSImage, "openebs-image", "eeacms/rsync:2.3", "The image used by OpenEBS disk free evaluation pod.")
	cmd.Flags().StringVar(&openEBSNode, "openebs-node", "", "Show OpenEBS free disk space only for provided node.")
	return cmd
}

func runCheck(*v1.StorageClass) (string, err){
	switch selectedClass.Provisioner {
	case openEBSLocalProvisioner:
		....

	default:
		return fmt.Errorf("provisioner %q is not supported", selectedClass.Provisioner)
	}
}

func findDefaultClass() (*v1.StorageClass, error) {
	var selectedClass *v1.StorageClass
	classes, err := clientSet.StorageV1().StorageClasses().List(cmd.Context(), metav1.ListOptions{})
	if err != nil {
		return nil, fmt.Errorf("failed to list cluster storage classes: %w", err)
	}

	for _, class := range classes.Items {
		val, ok := class.Annotations[isDefaultStorageClassAnnotation]
		if ok && val == "true" && forStorageClass == "" {
			selectedClass = &class
			logger.Printf("Selected default storage class %q", selectedClass.Name)
			break
		}

		if class.Name == forStorageClass {
			selectedClass = &class
			logger.Printf("Selected storage class %q", selectedClass.Name)
			break
		}
	}

	if selectedClass == nil {
		if forStorageClass == "" {
			return fmt.Errorf("failed to find default storage class")
		}
		return fmt.Errorf("failed to find storage class %q", forStorageClass)
	}
	return selectedClass,
}
```

WDYT? Is that make sense?

Reply:
The suggestion look for cohesion so we can grow in the future. 
This method has more than 30 lines which would not pass in a lint check for complexity for example. 

Anyway, I will not block the PR because of this.

Reply:
I agree with adding complexity when it is necessary and I don't consider that this is the case. Moreover 30 lines for go (where possibly half would be error treatment) is not a long function.

Reply:
As this is not a blocker I am resolving it.